File: src/main/scala/com/proyecto/api/chat/application/service/SendMessageService.scala
package com.proyecto.api.chat.application.service

import cats.effect.IO
import com.proyecto.api.chat.application.port.in.{SendMessageUseCase, SendMessageCommand}
import com.proyecto.api.chat.application.port.out.MessageRepository
import com.proyecto.api.chat.domain.model._

class SendMessageService(messageRepository: MessageRepository) extends SendMessageUseCase {

  override def execute(command: SendMessageCommand): IO[Long] = {
    for {
      createdAt <- IO.realTimeInstant
      
      content <- buildContent(command)
      
      // ID es 0 temporalmente, la BD lo ignorará y asignará uno real vía RETURNING
      message = Message(
        id = MessageId(0L), 
        roomId = RoomId(command.roomId),
        senderId = UserId(command.senderId),
        messageType = MessageType.fromCode(command.messageType),
        content = content,
        createdAt = createdAt,
        isRead = false
      )
      
      // El repositorio devuelve la entidad reconstruida con el ID generado real
      savedMessage <- messageRepository.save(message)
      
    } yield savedMessage.id.value
  }

  private def buildContent(cmd: SendMessageCommand): IO[MessageContent] = IO.delay {
    MessageType.fromCode(cmd.messageType) match {
      case MessageType.TEXT => 
        MessageContent.TextContent(cmd.text.getOrElse(throw new IllegalArgumentException("Text is required for TEXT")))
      case MessageType.AUDIO => 
        MessageContent.AudioContent(
          cmd.url.getOrElse(throw new IllegalArgumentException("URL is required")),
          cmd.durationSeconds.getOrElse(throw new IllegalArgumentException("Duration is required")),
          cmd.waveform
        )
      case MessageType.PHOTO =>
        MessageContent.PhotoContent(
          cmd.url.getOrElse(throw new IllegalArgumentException("URL is required")),
          cmd.sizeBytes,
          cmd.caption
        )
      case MessageType.VIDEO =>
        MessageContent.VideoContent(
          cmd.url.getOrElse(throw new IllegalArgumentException("URL is required")),
          cmd.durationSeconds.getOrElse(0.0),
          cmd.sizeBytes,
          cmd.thumb
        )
      case MessageType.STICKER => 
        MessageContent.StickerContent(cmd.stickerId.getOrElse(throw new IllegalArgumentException("StickerId is required")))
      case MessageType.DOCUMENT =>
        MessageContent.DocumentContent(
          cmd.url.getOrElse(throw new IllegalArgumentException("URL is required for DOCUMENT")),
          cmd.fileName.getOrElse(throw new IllegalArgumentException("fileName is required for DOCUMENT")),
          cmd.fileExtension.getOrElse(throw new IllegalArgumentException("fileExtension is required for DOCUMENT")),
          cmd.sizeBytes.getOrElse(throw new IllegalArgumentException("sizeBytes is required for DOCUMENT"))
        )
    }
  }
}

File: SphereChat.Api/Application/Services/SendMessageService.cs
using SphereChat.Api.Application.Ports.In;
using SphereChat.Api.Application.Ports.Out;
using SphereChat.Api.Domain.Models.Chat;

namespace SphereChat.Api.Application.Services;

/// <summary>
/// Servicio que orquesta el envío de mensajes.
/// Equivalente a SendMessageService.scala
/// </summary>
public class SendMessageService : ISendMessageUseCase
{
    private readonly IMessageRepository _messageRepository;

    public SendMessageService(IMessageRepository messageRepository)
    {
        _messageRepository = messageRepository;
    }

    public async Task<long> ExecuteAsync(SendMessageCommand command)
    {
        var content = BuildContent(command);

        // ID es 0 temporalmente, la BD lo asignará vía RETURNING
        var message = new Message(
            Id: 0L,
            RoomId: command.RoomId,
            SenderId: command.SenderId,
            MessageType: (MessageType)command.MessageType,
            Content: content,
            CreatedAt: DateTime.UtcNow,
            IsRead: false
        );

        var savedMessage = await _messageRepository.SaveAsync(message);
        return savedMessage.Id;
    }

    private static MessageContent BuildContent(SendMessageCommand cmd)
    {
        var msgType = (MessageType)cmd.MessageType;
        return msgType switch
        {
            MessageType.Text => new TextContent(
                cmd.Text ?? throw new ArgumentException("Text is required for TEXT")),

            MessageType.Audio => new AudioContent(
                cmd.Url ?? throw new ArgumentException("URL is required"),
                cmd.DurationSeconds ?? throw new ArgumentException("Duration is required"),
                cmd.Waveform),

            MessageType.Photo => new PhotoContent(
                cmd.Url ?? throw new ArgumentException("URL is required"),
                cmd.SizeBytes,
                cmd.Caption),

            MessageType.Video => new VideoContent(
                cmd.Url ?? throw new ArgumentException("URL is required"),
                cmd.DurationSeconds ?? 0.0,
                cmd.SizeBytes,
                cmd.Thumb),

            MessageType.Sticker => new StickerContent(
                StickerId: cmd.StickerId ?? throw new ArgumentException("StickerId requerido"),
                IsAnimated: cmd.IsAnimated
            ),

            MessageType.Document => new DocumentContent(
                cmd.Url ?? throw new ArgumentException("URL is required for DOCUMENT"),
                cmd.FileName ?? throw new ArgumentException("fileName is required for DOCUMENT"),
                cmd.FileExtension ?? throw new ArgumentException("fileExtension is required for DOCUMENT"),
                cmd.SizeBytes ?? throw new ArgumentException("sizeBytes is required for DOCUMENT")),

            _ => throw new ArgumentException($"MessageType desconocido: {cmd.MessageType}")
        };
    }
}
